panic when a peer is temporarily unavailable

We've seen multiple instances of an MT panicking when it got queried while some of its peers were temporarily unavailable. This is one of the messages I saw:

```
2020-03-31 21:30:57.829 [INFO] Proxying to Graphite because of error: unknown function "groupByNode"
2020/03/31 21:30:57 [WARN] memberlist: Was able to connect to <pod name> but other probes failed, network may be misconfigured
2020/03/31 21:30:58 [DEBUG] memberlist: Failed ping: <pod name> (timeout reached)
panic: runtime error: index out of range

goroutine 2765795 [running]:
github.com/grafana/metrictank/api.queryPeers.func1(0xc0317d6660, 0xc02898c480, 0x117d600, 0xc0025e80c0, 0xc0006d25a0, 0xc0006fc560)
        /go/src/github.com/grafana/metrictank/api/cluster.go:559 +0xd20
created by github.com/grafana/metrictank/api.queryPeers
        /go/src/github.com/grafana/metrictank/api/cluster.go:526 +0xb1
```

The running version was `v0.13.1-458-ged6257f`. 

Probably the code shouldn't assume that each group has at least one ready member here: https://github.com/grafana/metrictank/blob/bdd9d1e4de447e11633ba53ee557fbcb5ffe470e/api/cluster.go#L559

--- a/api/cluster.go
+++ b/api/cluster.go
@@ -563,6 +563,12 @@ func queryPeers(ctx context.Context, peerGroups map[int32][]cluster.Node, name s
 		}
 
 		for group, peers := range peerGroups {
+			if len(peers) == 0 {
+				errorChan <- fmt.Errorf("peer group %d has no peers", group)
+				delete(peerGroups, group)
+				continue
+			}
+
 			nextPeer := peers[0]
 			// shift nextPeer from the group
 			peerGroups[group] = peers[1:]

--- a/api/cluster.go
+++ b/api/cluster.go
@@ -564,7 +564,7 @@ func queryPeers(ctx context.Context, peerGroups map[int32][]cluster.Node, name s
 
 		for group, peers := range peerGroups {
 			if len(peers) == 0 {
-				errorChan <- fmt.Errorf("peer group %d has no peers", group)
+				log.Warningf("Peer group %d has no peers", group)
 				delete(peerGroups, group)
 				continue
 			}

--- a/api/cluster.go
+++ b/api/cluster.go
@@ -564,7 +564,7 @@ func queryPeers(ctx context.Context, peerGroups map[int32][]cluster.Node, name s
 
 		for group, peers := range peerGroups {
 			if len(peers) == 0 {
-				log.Warningf("Peer group %d has no peers", group)
+				log.Warningf("HTTP Peer group %d has no peers", group)
 				delete(peerGroups, group)
 				continue
 			}

--- a/api/dataprocessor.go
+++ b/api/dataprocessor.go
@@ -240,10 +240,12 @@ func (s *Server) getTargetsRemote(ctx context.Context, ss *models.StorageStats, 
 	shardReqs := make(map[int32][]models.Req)
 	for _, nodeReqs := range remoteReqs {
 		shardID := nodeReqs[0].Node.GetPartitions()[0]
-		if peers := allPeers[shardID]; len(peers) > 0 {
-			requiredPeers[shardID] = peers
-			shardReqs[shardID] = append(shardReqs[shardID], nodeReqs...)
+		peers := allPeers[shardID]
+		if len(peers) == 0 {
+			return nil, fmt.Errorf("Shard %d has gone unavailable before /getdata", shardID)
 		}
+		requiredPeers[shardID] = peers
+		shardReqs[shardID] = append(shardReqs[shardID], nodeReqs...)
 	}
 
 	rCtx, cancel := context.WithCancel(ctx)

--- a/api/dataprocessor.go
+++ b/api/dataprocessor.go
@@ -240,8 +240,10 @@ func (s *Server) getTargetsRemote(ctx context.Context, ss *models.StorageStats, 
 	shardReqs := make(map[int32][]models.Req)
 	for _, nodeReqs := range remoteReqs {
 		shardID := nodeReqs[0].Node.GetPartitions()[0]
-		requiredPeers[shardID] = allPeers[shardID]
-		shardReqs[shardID] = append(shardReqs[shardID], nodeReqs...)
+		if peers := allPeers[shardID]; len(peers) > 0 {
+			requiredPeers[shardID] = peers
+			shardReqs[shardID] = append(shardReqs[shardID], nodeReqs...)
+		}
 	}
 
 	rCtx, cancel := context.WithCancel(ctx)

